src/main/kotlin/org/rust/ide/intentions/DemorgansLawIntention.kt

```diff
@@ -0,0 +1,107 @@
+package org.rust.ide.intentions
+
+import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.tree.IElementType
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.util.elementType
+import org.rust.lang.core.psi.util.parentOfType
+
+class DemorgansLawIntention : PsiElementBaseIntentionAction() {
+    override fun getText() = "DeMorgan Law"
+    override fun getFamilyName() = text
+    override fun startInWriteAction() = true
+
+    private data class Context(
+        val binaryExpr: RustBinaryExprElement,
+        val binaryOpType: IElementType
+    )
+
+    private fun findContext(element: PsiElement): Context? {
+        if (!element.isWritable) return null
+
+        val binExpr = element.parentOfType<RustBinaryExprElement>() ?: return null
+        val opType = binExpr.operatorType
+        if (opType == RustTokenElementTypes.ANDAND || opType == RustTokenElementTypes.OROR) {
+            return Context(binExpr, opType)
+        }
+        return null
+    }
+
+    override fun invoke(project: Project, editor: Editor?, element: PsiElement) {
+        val ctx = findContext(element) ?: return
+        var binExpr = ctx.binaryExpr
+        val opType = ctx.binaryOpType
+
+        while (binExpr.parent is RustBinaryExprElement) {
+            binExpr = binExpr.parent as RustBinaryExprElement
+        }
```

Comment:
Rather than mutating `binExpr` I would introduce new immutable variable with more meaningful name: `topBinExpr` or sth like that.